crates/communication/src/server/databases/router.rs

```diff
@@ -0,0 +1,358 @@
+use std::collections::{hash_map::Entry, BTreeMap, HashMap};
+
+use tokio::{
+    spawn,
+    sync::mpsc::{Receiver, Sender},
+    task::JoinHandle,
+};
+
+use crate::server::messages::{
+    DatabaseRequest, Path, Response, TextualDatabaseResponse, TextualResponse, Type,
+};
+
+use super::{Client, ClientRequest, Request};
+
+pub fn router(mut request_receiver: Receiver<Request>) -> JoinHandle<()> {
+    spawn(async move {
+        let mut request_channels_of_cyclers = HashMap::new();
+        let mut cached_cycler_instances = HashMap::new();
+
+        while let Some(request) = request_receiver.recv().await {
+            match request {
+                Request::ClientRequest(request) => {
+                    forward_client_request_to_provider(
+                        request,
+                        &request_channels_of_cyclers,
+                        &mut cached_cycler_instances,
+                    )
+                    .await
+                }
+                Request::RegisterCycler {
+                    cycler_instance,
+                    fields,
+                    request_sender,
+                } => {
+                    request_channels_of_cyclers.insert(cycler_instance, (fields, request_sender));
+                }
+            }
+        }
+    })
+}
+
+async fn forward_client_request_to_provider(
+    request: ClientRequest,
+    request_channels_of_cyclers: &HashMap<String, (BTreeMap<Path, Type>, Sender<ClientRequest>)>,
+    cached_cycler_instances: &mut HashMap<(Client, usize), String>,
+) {
+    match &request.request {
+        DatabaseRequest::GetFields { id } => {
+            // TODO: directly generate BTreeMap in SerializeHierarchy
+            let _ = request
+                .client
+                .response_sender
+                .send(Response::Textual(TextualResponse::Databases(
+                    TextualDatabaseResponse::GetFields {
+                        id: *id,
+                        fields: request_channels_of_cyclers
+                            .iter()
+                            .map(|(cycler_instance, (fields, _request_sender))| {
+                                (cycler_instance.clone(), fields.clone())
+                            })
+                            .collect(),
+                    },
+                )))
+                .await;
+        }
+        DatabaseRequest::GetNext {
+            id,
+            cycler_instance,
+            ..
+        }
+        | DatabaseRequest::Subscribe {
+            id,
+            cycler_instance,
+            ..
+        } => {
+            if matches!(request.request, DatabaseRequest::Subscribe { .. }) {
+                cached_cycler_instances
+                    .insert((request.client.clone(), *id), cycler_instance.clone());
+            }
+
+            match request_channels_of_cyclers.get(cycler_instance) {
+                Some((_fields, request_channel)) => {
+                    let _ = request_channel.send(request).await;
```

Comment:
Jo, fixed.

Reply:
What about the other cases? I see four more in this file, five in the provider and one in the receiver.
All of those remaining ones send into the `response_sender` or `outputs_sender`, which are handled by the sender and router respectively. Since these two tasks keep running as long as there are senders for their channels, these operations should also never fail.

Reply:
Hmmm, am I correct that I can expect every single result where any channel is involved. I think, there are no corner cases, right? All receivers are not canceled/dropped and live as long as at least one sender exists. Should be safe to send at any time. Could you check for situations where `keep_running` or `keep_only_self_running` are canceled or any errors would lead to a receiver drop? 🙏

Reply:
The five send calls in the provider are directed at the client's response sender. The corresponding receiver is owned by the receiver task, which doesn't terminate upon token cancellation.

The send call in the receiver (unsubscribe all upon termination) is directed at the outputs_sender. The corresponding receiver is owned by the outputs router which also does not terminate upon token cancellation.

Reply:
Expecting all channel sends now.